### PR TITLE
Add podresourcesstore in awscontainerinsightreceiver

### PR DESCRIPTION
**Pull request by aditya-purang**
### Description:
    - Add a new store in awscontainerinsightreceiver, this store gets the mapping between pod/container and their allocated devices.
    PS : we need to add the mount volume mount and volume for pod-resources in the agent operator package, will ask @ssamartp to add those in his PR.
    

### Testing:
Tested the api in a test cluster with a custom script:
```
&PodResources{
  Name: trn1-mlp,
  Namespace: default,
  Containers: []*ContainerResources{
    &ContainerResources{
      Name: trn1-mlp,
      Devices: []*ContainerDevices{
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [9], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [8], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [1], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [4], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [0], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [11], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [3], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [7], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [12], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [6], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [13], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [2], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [10], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [14], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [5], Topology: nil},
        &ContainerDevices{ResourceName: aws.amazon.com/neuron, DeviceIds: [15], Topology: nil},
      },
      CpuIds: [],
      Memory: []*ContainerMemory{},
      DynamicResources: []*DynamicResource{},
    },
  },
}

```

also tested on a the test cluster and printed the map, some of the entries in the map: 
```
2024-03-01T17:06:36Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_2_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:36Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_14_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:36Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_5_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:36Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_9_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:36Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_12_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_14_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_4_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_7_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_3_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_8_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_1_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_5_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
2024-03-01T17:06:46Z I! {"caller":"stores/podresourcesstore.go:146","msg":"/nContainerInfo : {default_trn1-mlp_trn1-mlp} -> ResourceInfo : {aws.amazon.com/neuron_10_}","kind":"receiver","name":"awscontainerinsightreceiver","data_type":"metrics"}
```

### CODE REVIEW

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+func NewPodResourcesClient() (*PodResourcesClient, error) {
+	podResourcesClient := &PodResourcesClient{}
+
+	conn, cleanup, err := podResourcesClient.connectToServer(socketPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to server: %w", err)
+	}
+	defer cleanup()
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:34]: Have you tested this?  I'm pretty sure you don't want to clean up the connection after finishing setup here - the connection needs to be active for you to call ListPods.  Cleanup should happen in a PodResourcesClient Shutdown() method.

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+			d := net.Dialer{}
+			return d.DialContext(ctx, "unix", addr)
+		}),
+	)
+
+	if err != nil {
+		return nil, func() {}, fmt.Errorf("failure connecting to '%s': %w", socket, err)
+	}
+
+	return conn, func() { conn.Close() }, nil
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:59]: Minor: if you end up saving `conn` as a field on the PodResourcesClient so you can clean it up later, then I think it's unnecessary complexity to return a cleanup wrapper function.  `Shutdown()` would simply call `conn.Close()` if `conn` is not nil.

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+func (p *PodResourcesClient) ListPods() (*podresourcesapi.ListPodResourcesResponse, error) {
+	_, err := os.Stat(socketPath)
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("socket path does not exist: %s", socketPath)
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to check socket path: %w", err)
+	}
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:68]: This should go in `connectToServer` I think (you don't even use `socketPath` in this method).

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+	"go.uber.org/zap"
+	v1 "k8s.io/kubelet/pkg/apis/podresources/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/stores/kubeletutil"
+)
+
+const (
+	socketPath        = "/var/lib/kubelet/pod-resources/kubelet.sock"
+	connectionTimeout = 10 * time.Second
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:20]: Not used in this class.

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+type ResourceInfo struct {
+	resourceName string
+	deviceID     string
+}
+
+type PodResourcesClientInterface interface {
+	ListPods() (*v1.ListPodResourcesResponse, error)
+}
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:42]: Should this go in podresourcesclient.go?

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+				containerInfo := ContainerInfo{
+					podName:       pod.GetName(),
+					namespace:     pod.GetNamespace(),
+					containerName: container.GetName(),
+				}
+
+				for _, deviceID := range device.GetDeviceIds() {
+					resourceInfo := ResourceInfo{
+						resourceName: device.GetResourceName(),
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:123]: We might want to pass in a set of device resource names that we are interested in, so that we don't waste space storing *all* devices.

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+type PodResourcesClientInterface interface {
+	ListPods() (*v1.ListPodResourcesResponse, error)
+}
+
+type PodResourcesStore struct {
+	containerInfoToResourcesMap map[ContainerInfo][]ResourceInfo
+	resourceToPodContainerMap   map[ResourceInfo]ContainerInfo
+	lastRefreshed               time.Time
+	ctx                         context.Context
+	cancel                      context.CancelFunc
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:49]: You should have a `Shutdown()` function that calls `cancel`.

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore_test.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+func (m *MockPodResourcesClient) ListPods() (*podresourcesv1.ListPodResourcesResponse, error) {
+	mockResp := &podresourcesv1.ListPodResourcesResponse{
+		PodResources: []*podresourcesv1.PodResources{
+			{
+				Name:      "test-pod",
+				Namespace: "test-namespace",
+				Containers: []*podresourcesv1.ContainerResources{
+					{
+						Name: "test-container",
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore_test.go:82]: Can you also add a container with no devices to test that nothing is stored for that one.

And also a test case with a completely empty response to make sure the code doesn't break.

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore_test.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+	store := &PodResourcesStore{
+		containerInfoToResourcesMap: make(map[ContainerInfo][]ResourceInfo),
+		resourceToPodContainerMap:   make(map[ResourceInfo]ContainerInfo),
+		lastRefreshed:               time.Now(),
+		ctx:                         context.Background(),
+		cancel:                      func() {},
+		logger:                      logger,
+		podResourcesClient:          &MockPodResourcesClient{},
+	}
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore_test.go:116]: Extract this to a helper method that you can call from each of the test cases.

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go (hunk 3 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:68]: yeah makes sense

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go (hunk 1 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:34]: hmm I see what you mean, will try to hack it onto my test cluster after all the changes you suggested

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go (hunk 2 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:59]: yeah, didn't know about the shutdown pattern but I'd definitely make things a lot more cleaner

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go (hunk 4 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:20]: removed

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go (hunk 6 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:123]: would it make sense to have a add_resource_name method which adds that to a set which we can then use for lookup and populate the map in the next run?

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go (hunk 7 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:49]: ack

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go (hunk 5 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:42]: had this just to make testing easier, this is similar to how the podstore tests things

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go (hunk 4 above)
Comment (straussb) [reply receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go]: `connectionTimeout` is not used either.

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go (hunk 6 above)
Comment (straussb) [reply receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:123]: Yes I like that.

File: receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+					namespace:     pod.GetNamespace(),
+					containerName: container.GetName(),
+				}
+
+				for _, deviceID := range device.GetDeviceIds() {
+					resourceInfo := ResourceInfo{
+						resourceName: device.GetResourceName(),
+						deviceID:     deviceID,
+					}
+					_, found := p.resourceNameSet[resourceInfo.resourceName]
```
[the remark this answers is not in the document]
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/podresourcesstore.go:132]: nice catch, updated the PR

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+	return resp, nil
+}
+
+func (p *PodResourcesClient) Shutdown() {
+	err := p.conn.Close()
+	if err != nil {
+		return
+	}
```
Comment (straussb) [inline receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:86]: Is this just to avoid the compiler giving you a warning about ignoring the error?

**Review by straussb — APPROVED** [review]

File: receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go (hunk 11 above)
Comment (aditya-purang) [reply receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/podresourcesclient.go:86]: yup

**Review by straussb — APPROVED** [review]

**Review by straussb — APPROVED** [review]

**Comment by movence** [comment]
Please update the target branch to `aws-cwa-dev`
